File: backend/app/schemas/progress.py
import uuid
from datetime import date

from pydantic import BaseModel


class DailyPoint(BaseModel):
    date: date
    tasks_completed: int
    score: float


class WeeklyPoint(BaseModel):
    week_start: date
    tasks_completed: int
    score: float


class CategoryBreakdownItem(BaseModel):
    category_id: uuid.UUID | None
    category_name: str
    color: str | None
    task_count: int
    score: float


class SnapshotRead(BaseModel):
    period_type: str
    period_start: date
    tasks_created: int
    tasks_completed: int
    total_score: float
    completion_rate: float
    avg_completion_time_hours: float | None

File: frontend/lib/progress.ts
import { apiFetch } from "@/lib/api";

export interface DailyPoint {
  date: string;
  tasks_completed: number;
  score: number;
}

export interface WeeklyPoint {
  week_start: string;
  tasks_completed: number;
  score: number;
}

export interface CategoryBreakdownItem {
  category_id: string | null;
  category_name: string;
  color: string | null;
  task_count: number;
  score: number;
}

export interface SnapshotRead {
  period_type: "week" | "month";
  period_start: string;
  tasks_created: number;
  tasks_completed: number;
  total_score: number;
  completion_rate: number;
  avg_completion_time_hours: number | null;
}

export function getDailySeries(days: number) {
  return apiFetch<DailyPoint[]>(`/api/v1/progress/daily?days=${days}`);
}

export function getWeeklySeries(weeks: number) {
  return apiFetch<WeeklyPoint[]>(`/api/v1/progress/weekly?weeks=${weeks}`);
}

export function getCategoryBreakdown() {
  return apiFetch<CategoryBreakdownItem[]>("/api/v1/progress/category-breakdown");
}

export function getSnapshot(periodType: "week" | "month") {
  return apiFetch<SnapshotRead>(`/api/v1/progress/snapshot?period_type=${periodType}`);
}
